File: Simple_NN_library/python/functions/loss_functions/classification.py
from Simple_NN_library.python.further_maths import*

class Classification:

    def get_func_name(self, func):
        self.list_of_functions=[self.binary_cross_entropy, self.negative_log_likelihood]
        func_name="unknown"
        if func==self.list_of_functions[0]:func_name="binary cross entropy"
        elif func==self.list_of_functions[1]:func_name="negative log likelihood"
        return func_name

    def get_func(self, name):
        self.list_of_functions=[self.binary_cross_entropy, self.negative_log_likelihood]
        if name=="binary cross entropy":return self.list_of_functions[0]
        elif name=="negative log likelihood":return self.list_of_functions[1]

    def binary_cross_entropy(self, output, expected_output):
        x=output
        z=expected_output

        result=0

        for i in range(len(x)):
            result+=x[i]*log(z[i])+(1-x[i])*log(1-z[i])
        return -result

    def negative_log_likelihood(self, output, expected_output):
        result=0
        for i in expected_output:
            result+=-log(i)
        return result

File: Simple_NN_library/C++/production_versioon/functions/loss_functions/classification.cpp
#include "classification.h"

using namespace Simple_NN_library;
using namespace std;
using namespace functions;
using namespace loss_functions;
using namespace further_maths;

string Classification::get_func_name(float func(float*, float*)){
  string func_name="unknown";
  if(*func==binary_cross_entropy){func_name="binary cross entropy";}
  else if(*func==negative_log_likelihood){func_name="negative log likelihood";}
  return func_name;
}
Classification::pointer Classification::get_func(string name){
  if(name=="binary cross entropy"){return binary_cross_entropy;}
  else if(name=="negative log likelihood"){return negative_log_likelihood;}
}

float Classification::binary_cross_entropy(float* output, float* expected_output){
  //Algorithm=-sum_i(x[i]*log(z[i])+(1-x[i])*log(1-z[i]))
  //x=output z=expected
  float* x=output;
  float* z=expected_output;

  float result=0;

  for(int i=0; i<sizeof(x); i++){
    result+=x[i]*log(z[i])+(1-x[i])*log(1-z[i]);
  }
  return -result;
}
float Classification::negative_log_likelihood(float* output, float* expected_output){
  //Is this right?
  float result=0;
  for(int i=0; i<sizeof(expected_output); i++){
    result+=-log(expected_output[i]);
  }
  return result;
}

/*
float Classification::margin_classifier(float output[], float expected_output[]){

}
float Classification::soft_margin_classifier(float output[], float expected_output[]){

}
*/
